scripts/find-unknown-comment-words

```diff
@@ -0,0 +1,143 @@
+#!/bin/bash
+#
+# Locate unknown words in C/C++ comments.  Uses "extract-comments"
+# and "ablexicon" scripts.
+#
+set -o nounset
+set -o pipefail
+set -o errexit
+set -f
+
+BLUE="\e[1;34m"
+GREEN="\e[1;32m"
+DEFAULTFG="\e[39m"
+
+function usage () {
+    echo "Find unknown words in C/C++ comments"
+    echo ""
+    echo "Usage:"
+    echo " ${0##*/} [options]"
+    echo ""
+    echo "Options:"
+    echo " -d, --directory    directory to scan (defaults to .)"
+    echo " -l, --lexicon      lexicon file (one word per line, default 'lexicon.txt')"
+    echo " -t, --terse        terse output only (enabled if no lexicon available)"
+    echo " -h, --help         display this help"
+    exit 1
+}
+
+#
+# Verify that required commands are present
+#
+REQUIRED=( "extract-comments" "ablexicon" "getopt" )
+for i in "${REQUIRED[@]}"
+do
+    command -v $i"" >/dev/null
+    if [ $? -ne "0" ]
+    then
+        echo "Can't find '"$i"' , exiting...">&2
+        exit 1
+    fi
+done
+
+GETOPT_OUT=`getopt -o htd:l: --long help,terse,directory:,lexicon: -n "${0##*/}" -- "$@"`
+if [ $? != 0 ]
+then
+    echo "Exiting..." >&2
+    exit 1
+fi
+
+eval set -- "$GETOPT_OUT"
+
+DIRNAME=/dev/fd/0
+LEXICON=
+STATUS=
+TERSE=
+while true; do
+  case "$1" in
+    -h | --help ) usage $0 ;;
+    -t | --terse ) TERSE=1; shift ;;
+    -d | --directory ) DIRNAME="$2"; shift 2 ;;
+    -l | --lexicon ) LEXICON="$2"; shift 2 ;;
+    -- ) shift; break ;;
+    * ) break ;;
+  esac
+done
+
+if [ ! -d $DIRNAME"" ]
+then
+    echo "Invalid directory: "$DIRNAME
+    usage
+fi
+
+if [ $LEXICON"" = "" ]
+then
+    if [ -f $DIRNAME/lexicon.txt ]
+    then
+        LEXICON=$DIRNAME/lexicon.txt
+    else
+        LEXICON=/dev/null
+        TERSE=1
+    fi
+fi
+
+TMPFILE=${0##*/}-$USER-$RANDOM
+unknowns=( "not-used" )     # get around empty array with nounset
+extract-comments `find $DIRNAME -name \*.[ch]` |
+    tr [:upper:] [:lower:] |
+    grep -o -E '[a-zA-Z]+' |
+    ablexicon -l $LEXICON > $TMPFILE
+readarray -O 1 -t unknowns < $TMPFILE
```

Comment:
Suggestion for future tinkering: If all this does is append a list of newline-separated words into an array, then I think `unknowns+=( $(cat $TMPFILE | tr "\n" " ") )` might be equivalent to the `readarray`, and does not require bash >= 4.

Reply:
You don't need the cat or the tr, since bash will split on $IFS.

unknowns+=( $(< $TMPFILE) )